Bug: https://trac.sagemath.org/ticket/32037

From ef69307c97c81a990182cf26246cdc1d87fdac6d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fr=C3=A9d=C3=A9ric=20Chapoton?= <chapoton@unistra.fr>
Date: Thu, 24 Jun 2021 09:16:25 +0200
Subject: first steps towards lcalc 2.0.0

---
 build/pkgs/lcalc/checksums.ini       | 8 ++++----
 build/pkgs/lcalc/package-version.txt | 2 +-
 src/sage/libs/lcalc/lcalc_sage.h     | 3 +--
 3 files changed, 6 insertions(+), 7 deletions(-)

From 8b5bb1adb2f163d2dc6bffc5ecbc3e1905faa30a Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sun, 27 Jun 2021 22:11:41 -0400
Subject: Trac #32037: update libs/lcalc API for lcalc-2.x.

The find_zeros_via_N_v() function was removed from the libLfunction
API in lcalc-2.0.0, and has been replaced by find_zeros(). This commit
updates sage/libs/lcalc/lcalc_Lfunction.pxd with the new function
name, arguments, and return type. The relevant bits can be found in
the lcalc/L.h header.
---
 src/sage/libs/lcalc/lcalc_Lfunction.pxd | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

From 3a746e933e8fbaed086d6c9ad991cdf842fd00d7 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sun, 27 Jun 2021 22:23:27 -0400
Subject: Trac #32037: use lcalc-2.x API in libs/lcalc/lcalc_Lfunction.pyx.

Now that we've declared the lcalc-2.x API in our pxd file, this commit
replaces most of the internal uses of find_zeros_via_N_v() with
find_zeros(). The one remaining instance within the Python
find_zeros_via_N() method was left alone, since most of that method
needs to be rewritten anyway.
---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

From 6fe69729900068a1dd4f546c2455cb6496343a92 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sun, 27 Jun 2021 22:44:04 -0400
Subject: Trac #32037: update find_zeros_via_N() to match the lcalc-2.x API.

Our python find_zeros_via_N() method previously delegated to the C++
find_zeros_via_N_v() method in libLfunction. That method has been
replaced, and the new one has a different set of arguments and some
new behavior. Here we update the arguments and documentation for
find_zeros_via_N() in SageMath to better match the underlying library.

The implementation of find_zeros_via_N() has also been changed, to
more directly call the library API using the arguments passed to
find_zeros_via_N(). It's not clear to me why the previous
incarnation messed around with MPFR so much.
---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 35 ++++++++++++++++-----------------
 1 file changed, 17 insertions(+), 18 deletions(-)

From ebf11dab2dda8c12252fa201d4278dd8b9a57739 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Fri, 2 Jul 2021 19:07:39 -0400
Subject: Trac 32037: fixup lcalc header name in sage/libs/lcalc/lcalc_sage.h.

---
 src/sage/libs/lcalc/lcalc_sage.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

From 67566b07cb21821799d99d4e3c2774e4e3a43c2c Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sat, 3 Jul 2021 08:01:46 -0400
Subject: Trac #32037: update the doctest for the DirichletGroup(5)[1]
 L-function.

Due to a change in libLfunction's API, this L-function is now
automatically determined to be complex, which means that we search for
zeros below the real axis as well. As a result, the first zero (in
magnitude) is now negative. (In the past, it was ignored without
do_negative=True). We update the doctest to include it.
---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

From eaa4e21f3ce2f92b44debbdc26f0f14f7acaad58 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sat, 3 Jul 2021 08:06:01 -0400
Subject: Trac #32037: lower all lcalc tolerances to 1e-8.

We're using machine doubles, so the existing tolerances on the order
of 1e-15 were asking for trouble. This fixes some newly-failing
doctests with lcalc-2.0.1.
---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

From b168a0951dfb3fc5477d53c43e509d084966049f Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sat, 3 Jul 2021 08:08:44 -0400
Subject: Trac #32037: add abs_tol to an lcalc test to fix a new failure.

---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

From a7c66e654ca7dd7f3b8a85c60ea8a67571a13540 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sat, 3 Jul 2021 08:10:55 -0400
Subject: Trac #32037: update an lcalc doctest that counts zeros.

One of the newly-failing doctests with lcalc-2.0.1 was for the `__N()`
method, which is supposed to count zeros. We were expecting 3.170439...
but with lcalc-2.0.1, the output is roughly 4.0, which seems to me a
more likely number of zeros. Let's just expect that instead.
---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

From 143c85d7929b905dbbead85b14c8f0f8e8c697c0 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sat, 3 Jul 2021 08:31:52 -0400
Subject: Trac #32037: add tolerance to a few sage/lfunctions/lcalc.py
 doctests.

Some doctests in sage/lfunctions/lcalc.py are failing due to numerical
noise with the latest lcalc-2.0.1. We add "abs_tol" in a few places to
make everything okay.
---
 src/sage/lfunctions/lcalc.py | 78 ++++++++++++++++++++++++++++++++++++++------
 1 file changed, 68 insertions(+), 10 deletions(-)

From eb84c7b2480a50e76a5869f1bd173974fa2cf74f Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sat, 3 Jul 2021 08:36:17 -0400
Subject: Trac #32037: delete a failing doctest.

The one remaining doctest failure after upgrading to lcalc-2.0.1 is
with the output from,

  sage: chi = DirichletGroup(5)[1]
  sage: L = Lfunction_from_character(chi, type="complex")
  sage: L.hardy_z_function(0.4 + 0.3*I)

This was discussed also on Trac #24820, and it seems that neither the
old output nor the new output can be justified. In other words, we
don't know if there's a regression, or if anything needs to be
changed. For now, to proceed with the lcalc upgrade, I have simply
deleted the doctest
---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 2 --
 1 file changed, 2 deletions(-)

From 648e1317adc6a62987ae60fdad7f84286c5577e4 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sat, 3 Jul 2021 08:42:11 -0400
Subject: Trac #32037: replace some ellipses with abs_tol.

There are a few comparisons in libs/lcalc/lcalc_Lfunction.pyx that are
being made as strings (with ellipses) that can instead be made easily
with abs_tol. We switch them, to clean up this file a bit.
---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 37 +++++++++++++++++----------------
 1 file changed, 19 insertions(+), 18 deletions(-)

From f80d33502244ffb92d69367c6d0a1384c6601f01 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sat, 3 Jul 2021 17:25:37 -0400
Subject: Trac #32037: adjust doctest tolerance for lcalc-2.0.1 in
 lseries_ell.py.

---
 src/sage/schemes/elliptic_curves/lseries_ell.py | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

From dc5766c85b1c03f8d3cf7fe34967f8755f49292b Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sat, 3 Jul 2021 17:27:46 -0400
Subject: Trac #32037: weaken doctest tolerance for lcalc-2.0.1 in
 modular/dirichlet.py.

---
 src/sage/modular/dirichlet.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

From efabc7ef80a4f3f5277d1177f0bc9b4c0b01352f Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Sat, 3 Jul 2021 17:40:45 -0400
Subject: Trac #32037: tweak a doctest in
 elliptic_curves/ell_rational_field.py.

With lcalc-2.0.1, an exception gets thrown in this file with a
slightly different number than was shown before. We add some ellipses
to it to allow the test to pass with the newer lcalc.
---
 src/sage/schemes/elliptic_curves/ell_rational_field.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

From 8a2e9d8e1f49af1889908fdc266f86a973b9b646 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Wed, 7 Jul 2021 07:54:10 -0400
Subject: Trac #32037: add tolerance to a test in
 sage/lfunctions/zero_sums.pyx.

---
 src/sage/lfunctions/zero_sums.pyx | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

From 22efbb4807a20aa4ff10a8b6b2d17d9157778e86 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Thu, 25 Nov 2021 11:07:46 -0500
Subject: Trac #32037: don't use "abs tol" for complex numbers.

Our "abs tol" doctest test modifier doesn't really handle complex
numbers, and only works incidentally when both components are parsed
into equal-length lists of floats. And moreover, doctest decorators
aren't fully appropriate in EXAMPLES, because the user can't reproduce
them. In this commit we replace a bunch of "abs tol" examples in
the sage.lcalc.lcalc_Lfunction module with explicit ones.
---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 92 +++++++++++++++++----------------
 1 file changed, 48 insertions(+), 44 deletions(-)

From acf4d44c932de5ff89a52e831cf5d62e4a863596 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Thu, 25 Nov 2021 11:10:41 -0500
Subject: Trac #32037: whitespace cleanup in sage.libs.lcalc.lcalc_Lfunction.

---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

From 1a2dd14015b4900fab3bf8836f8929e43cd827cc Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Thu, 25 Nov 2021 11:27:03 -0500
Subject: Trac #32037: example cosmetology in sage.libs.lcalc.lcalc_Lfunction.

Just making some of our examples more readable and self-contained.
---
 src/sage/libs/lcalc/lcalc_Lfunction.pyx | 68 ++++++++++++++++++++++++---------
 1 file changed, 49 insertions(+), 19 deletions(-)



--- a/sage/src/sage/libs/lcalc/lcalc_sage.h
+++ b/sage/src/sage/libs/lcalc/lcalc_sage.h
@@ -1,4 +1,4 @@
-#include "Lfunction/L.h"
+#include "lcalc/L.h"
 int *new_ints(int l)
 {
     return new int[l];
@@ -62,4 +62,3 @@
     cout << "Value at 1"  << L->value(1.0) <<endl;
     cout << "Value at .5+I"  << L->value(.5+I) <<endl;
 }
-
--- a/sage/src/sage/libs/lcalc/lcalc_Lfunction.pxd
+++ b/sage/src/sage/libs/lcalc/lcalc_Lfunction.pxd
@@ -21,7 +21,7 @@
         int (* compute_rank) ()
         double (* N) (double T)
         void  (* find_zeros_v)(double T1, double T2, double stepsize, doublevec result )
-        void (*find_zeros_via_N_v)(long count,int do_negative,double max_refine, int rank, int test_explicit_formula, doublevec result)
+        int (*find_zeros)(long count, long start, double max_refine, int rank, const char* message_stamp, doublevec* result)
         void (*print_data_L)()
 
         #Constructor and destructor
@@ -38,7 +38,7 @@
         double (* N) (double T)
         double *dirichlet_coefficient
         void  (* find_zeros_v)(double T1, double T2, double stepsize, doublevec result )
-        void (*find_zeros_via_N_v)(long count,int do_negative,double max_refine, int rank, int test_explicit_formula, doublevec result)
+        int (*find_zeros)(long count, long start, double max_refine, int rank, const char* message_stamp, doublevec* result)
         void (*print_data_L)()
 
         #Constructor and destructor
@@ -54,7 +54,7 @@
         int (* compute_rank) ()
         double (* N) (double T)
         void  (* find_zeros_v)(double T1, double T2, double stepsize, doublevec result )
-        void (*find_zeros_via_N_v)(long count,int do_negative,double max_refine, int rank, int test_explicit_formula, doublevec result)
+        int (*find_zeros)(long count, long start, double max_refine, int rank, const char* message_stamp, doublevec* result)
         void (*print_data_L)()
 
         #Constructor and destructor
@@ -70,7 +70,7 @@
         int (* compute_rank) ()
         double (* N) (double T)
         void  (* find_zeros_v)(double T1, double T2, double stepsize, doublevec result )
-        void (*find_zeros_via_N_v)(long count,int do_negative,double max_refine, int rank, int test_explicit_formula, doublevec result)#puts result in vector<double> result
+        int (*find_zeros)(long count, long start, double max_refine, int rank, const char* message_stamp, doublevec* result)
         void (*find_zeros_via_N)(long count,int do_negative,double max_refine, int rank, int test_explicit_formula, char *filename) #puts result in filename
 
         #Constructor and destructor
@@ -111,7 +111,7 @@
     #strange bug, replacing Double with double gives me a compile error
     cdef Double __typedN(self, double T)
     cdef void __find_zeros_v(self, double T1, double T2, double stepsize,doublevec *result)
-    cdef void __find_zeros_via_N_v(self, long count,int do_negative,double max_refine, int rank, int test_explicit_formula,doublevec *result)
+    cdef int __find_zeros(self, long count, long start, double max_refine, int rank, const char* message_stamp, doublevec* result)
 
     cdef str _repr
 
--- a/sage/src/sage/libs/lcalc/lcalc_Lfunction.pyx
+++ b/sage/src/sage/libs/lcalc/lcalc_Lfunction.pyx
@@ -140,32 +140,43 @@
 
         EXAMPLES::
 
-            sage: chi = DirichletGroup(5)[2] #This is a quadratic character
             sage: from sage.libs.lcalc.lcalc_Lfunction import *
-            sage: L=Lfunction_from_character(chi, type="int")
-            sage: L.value(.5)  # abs tol 3e-15
-            0.231750947504016 + 5.75329642226136e-18*I
-            sage: L.value(.2+.4*I)
-            0.102558603193... + 0.190840777924...*I
+            sage: chi = DirichletGroup(5)[2] # This is a quadratic character
+            sage: L = Lfunction_from_character(chi, type="int")
+            sage: (L.value(0.5) - 0.231750947504016).abs() < 1e-8
+            True
+            sage: v = L.value(0.2 + 0.4*I)
+            sage: (v - (0.102558603193 + 0.190840777924*I)).abs() < 1e-8
+            True
+            sage: L = Lfunction_from_character(chi, type="double")
+            sage: (L.value(0.6) - 0.274633355856345).abs() < 1e-8
+            True
+            sage: v = L.value(0.6 + I)
+            sage: (v - (0.362258705721 + 0.43388825062*I)).abs() < 1e-8
+            True
 
-            sage: L=Lfunction_from_character(chi, type="double")
-            sage: L.value(.6)  # abs tol 3e-15
-            0.274633355856345 + 6.59869267328199e-18*I
-            sage: L.value(.6+I)
-            0.362258705721... + 0.433888250620...*I
+        ::
 
+            sage: from sage.libs.lcalc.lcalc_Lfunction import *
             sage: chi = DirichletGroup(5)[1]
-            sage: L=Lfunction_from_character(chi, type="complex")
-            sage: L.value(.5)
-            0.763747880117... + 0.216964767518...*I
-            sage: L.value(.6+5*I)
-            0.702723260619... - 1.10178575243...*I
-
-            sage: L=Lfunction_Zeta()
-            sage: L.value(.5)
-            -1.46035450880...
-            sage: L.value(.4+.5*I)
-            -0.450728958517... - 0.780511403019...*I
+            sage: L = Lfunction_from_character(chi, type="complex")
+            sage: v = L.value(0.5)
+            sage: (v - (0.763747880117 + 0.21696476751*I)).abs() < 1e-8
+            True
+            sage: v = L.value(0.6 + 5*I)
+            sage: (v - (0.702723260619 - 1.10178575243*I)).abs() < 1e-8
+            True
+
+        ::
+
+            sage: from sage.libs.lcalc.lcalc_Lfunction import *
+            sage: L = Lfunction_Zeta()
+            sage: (L.value(0.5) + 1.46035450880).abs() < 1e-8
+            True
+            sage: v = L.value(0.4 + 0.5*I)
+            sage: (v - (-0.450728958517 - 0.780511403019*I)).abs() < 1e-8
+            True
+
         """
         cdef ComplexNumber complexified_s = CCC(s)
         cdef c_Complex z = new_Complex(mpfr_get_d(complexified_s.__re, MPFR_RNDN), mpfr_get_d(complexified_s.__im, MPFR_RNDN))
@@ -175,34 +186,39 @@
     def hardy_z_function(self, s):
         """
         Computes the Hardy Z-function of the L-function at s
-        
+
         INPUT:
 
         - ``s`` - a complex number with imaginary part between -0.5 and 0.5
 
         EXAMPLES::
 
-            sage: chi = DirichletGroup(5)[2]  # Quadratic character
             sage: from sage.libs.lcalc.lcalc_Lfunction import *
+            sage: chi = DirichletGroup(5)[2]  # Quadratic character
             sage: L = Lfunction_from_character(chi, type="int")
-            sage: L.hardy_z_function(0)
-            0.231750947504... 
-            sage: L.hardy_z_function(.5).imag()  # abs tol 1e-15
-            1.17253174178320e-17
-            sage: L.hardy_z_function(.4+.3*I)
-            0.2166144222685... - 0.00408187127850...*I
+            sage: (L.hardy_z_function(0) - 0.231750947504).abs() < 1e-8
+            True
+            sage: L.hardy_z_function(0.5).imag().abs() < 1e-8
+            True
+
+        ::
+
+            sage: from sage.libs.lcalc.lcalc_Lfunction import *
             sage: chi = DirichletGroup(5)[1]
             sage: L = Lfunction_from_character(chi, type="complex")
-            sage: L.hardy_z_function(0)
-            0.793967590477...
-            sage: L.hardy_z_function(.5).imag()  # abs tol 1e-15
-            0.000000000000000
+            sage: (L.hardy_z_function(0) - 0.793967590477).abs() < 1e-8
+            True
+            sage: L.hardy_z_function(0.5).imag().abs() < 1e-8
+            True
+
+        ::
+
+            sage: from sage.libs.lcalc.lcalc_Lfunction import *
             sage: E = EllipticCurve([-82,0])
             sage: L = Lfunction_from_elliptic_curve(E, number_of_coeffs=40000)
-            sage: L.hardy_z_function(2.1)
-            -0.00643179176869...
-            sage: L.hardy_z_function(2.1).imag()  # abs tol 1e-15
-            -3.93833660115668e-19
+            sage: (L.hardy_z_function(2.1) - (-0.006431791768)).abs() < 1e-8
+            True
+
         """
         #This takes s -> .5 + I*s
         cdef ComplexNumber complexified_s = CCC(0.5)+ CCC(0,1)*CCC(s)
@@ -218,15 +234,20 @@
 
         EXAMPLES::
 
-            sage: chi = DirichletGroup(5)[2] #This is a quadratic character
             sage: from sage.libs.lcalc.lcalc_Lfunction import *
-            sage: L=Lfunction_from_character(chi, type="int")
+            sage: chi = DirichletGroup(5)[2] # This is a quadratic character
+            sage: L = Lfunction_from_character(chi, type="int")
             sage: L.compute_rank()
             0
-            sage: E=EllipticCurve([-82,0])
-            sage: L=Lfunction_from_elliptic_curve(E, number_of_coeffs=40000)
+
+        ::
+
+            sage: from sage.libs.lcalc.lcalc_Lfunction import *
+            sage: E = EllipticCurve([-82,0])
+            sage: L = Lfunction_from_elliptic_curve(E, number_of_coeffs=40000)
             sage: L.compute_rank()
             3
+
         """
         return self.__compute_rank()
 
@@ -241,8 +262,8 @@
             sage: from sage.libs.lcalc.lcalc_Lfunction import *
             sage: chi = DirichletGroup(5)[2] #This is a quadratic character
             sage: L=Lfunction_from_character(chi, type="complex")
-            sage: L.__N(10)
-            3.17043978326...
+            sage: L.__N(10) # abs tol 1e-8
+            4.0
         """
         cdef RealNumber real_T=RRR(T)
         cdef double double_T = mpfr_get_d(real_T.value, MPFR_RNDN)
@@ -260,7 +281,7 @@
         Use :meth:`find_zeros_via_N` for slower but more rigorous computation.
 
         INPUT:
-        
+
         - ``T1`` -- a real number giving the lower bound
         - ``T2`` -- a real number giving the upper bound
         - ``stepsize`` -- step size to be used for the zero search
@@ -272,21 +293,26 @@
         EXAMPLES::
 
             sage: from sage.libs.lcalc.lcalc_Lfunction import *
-            sage: chi = DirichletGroup(5)[2] #This is a quadratic character
-            sage: L=Lfunction_from_character(chi, type="int")
+            sage: chi = DirichletGroup(5)[2] # This is a quadratic character
+            sage: L = Lfunction_from_character(chi, type="int")
             sage: L.find_zeros(5,15,.1)
             [6.64845334472..., 9.83144443288..., 11.9588456260...]
-
-            sage: L=Lfunction_from_character(chi, type="double")
+            sage: L = Lfunction_from_character(chi, type="double")
             sage: L.find_zeros(1,15,.1)
             [6.64845334472..., 9.83144443288..., 11.9588456260...]
 
+        ::
+
+            sage: from sage.libs.lcalc.lcalc_Lfunction import *
             sage: chi = DirichletGroup(5)[1]
-            sage: L=Lfunction_from_character(chi, type="complex")
+            sage: L = Lfunction_from_character(chi, type="complex")
             sage: L.find_zeros(-8,8,.1)
             [-4.13290370521..., 6.18357819545...]
 
-            sage: L=Lfunction_Zeta()
+        ::
+
+            sage: from sage.libs.lcalc.lcalc_Lfunction import *
+            sage: L = Lfunction_Zeta()
             sage: L.find_zeros(10,29.1,.1)
             [14.1347251417..., 21.0220396387..., 25.0108575801...]
         """
@@ -307,18 +333,21 @@
         return returnvalue
 
     #The default values are from L.h. See L.h
-    def find_zeros_via_N(self, count=0, do_negative=False, max_refine=1025,
-                         rank=-1, test_explicit_formula=0):
+    def find_zeros_via_N(self, count=0, start=0, max_refine=1025, rank=-1):
         """
-        Finds ``count`` number of zeros with positive imaginary part
-        starting at real axis. This function also verifies that all
-        the zeros have been found.
+        Find ``count`` zeros (in order of increasing magnitude) and output
+        their imaginary parts. This function verifies that no zeros
+        are missed, and that all values output are indeed zeros.
+
+        If this L-function is self-dual (if its Dirichlet coefficients
+        are real, up to a tolerance of 1e-6), then only the zeros with
+        positive imaginary parts are output. Their conjugates, which
+        are also zeros, are not output.
 
         INPUT:
 
         - ``count`` - number of zeros to be found
-        - ``do_negative`` - (default: False) False to ignore zeros below the
-          real axis.
+        - ``start`` - (default: 0) how many initial zeros to skip
         - ``max_refine`` - when some zeros are found to be missing, the step
           size used to find zeros is refined. max_refine gives an upper limit
           on when lcalc should give up. Use default value unless you know
@@ -326,44 +355,51 @@
         - ``rank`` - integer (default: -1) analytic rank of the L-function.
           If -1 is passed, then we attempt to compute it. (Use default if in
           doubt)
-        - ``test_explicit_formula`` - integer (default: 0) If nonzero, test
-          the explicit formula for additional confidence that all the zeros
-          have been found and are accurate. This is still being tested, so
-          using the default is recommended.
 
         OUTPUT:
-        
+
         list -- A list of the imaginary parts of the zeros that have been found
 
         EXAMPLES::
 
             sage: from sage.libs.lcalc.lcalc_Lfunction import *
             sage: chi = DirichletGroup(5)[2] #This is a quadratic character
-            sage: L=Lfunction_from_character(chi, type="int")
+            sage: L = Lfunction_from_character(chi, type="int")
             sage: L.find_zeros_via_N(3)
             [6.64845334472..., 9.83144443288..., 11.9588456260...]
-
-            sage: L=Lfunction_from_character(chi, type="double")
+            sage: L = Lfunction_from_character(chi, type="double")
             sage: L.find_zeros_via_N(3)
             [6.64845334472..., 9.83144443288..., 11.9588456260...]
 
+        ::
+
+            sage: from sage.libs.lcalc.lcalc_Lfunction import *
             sage: chi = DirichletGroup(5)[1]
-            sage: L=Lfunction_from_character(chi, type="complex")
-            sage: L.find_zeros_via_N(3)
-            [6.18357819545..., 8.45722917442..., 12.6749464170...]
+            sage: L = Lfunction_from_character(chi, type="complex")
+            sage: zeros = L.find_zeros_via_N(3)
+            sage: (zeros[0] - (-4.13290370521286)).abs() < 1e-8
+            True
+            sage: (zeros[1]  - 6.18357819545086).abs() < 1e-8
+            True
+            sage: (zeros[2]  - 8.45722917442320).abs() < 1e-8
+            True
+
+        ::
 
-            sage: L=Lfunction_Zeta()
+            sage: from sage.libs.lcalc.lcalc_Lfunction import *
+            sage: L = Lfunction_Zeta()
             sage: L.find_zeros_via_N(3)
             [14.1347251417..., 21.0220396387..., 25.0108575801...]
+
         """
-        cdef Integer count_I = Integer(count)
-        cdef Integer do_negative_I = Integer(do_negative)
-        cdef RealNumber max_refine_R = RRR(max_refine)
-        cdef Integer rank_I = Integer(rank)
-        cdef Integer test_explicit_I = Integer(test_explicit_formula)
+
+        # This is the default value for message_stamp, but we have to
+        # pass it explicitly since we're passing in the next argument,
+        # our &result pointer.
+        cdef const char* message_stamp = ""
         cdef doublevec result
         sig_on()
-        self.__find_zeros_via_N_v(mpz_get_si(count_I.value), mpz_get_si(do_negative_I.value), mpfr_get_d(max_refine_R.value, MPFR_RNDN), mpz_get_si(rank_I.value), mpz_get_si(test_explicit_I.value), &result)
+        self.__find_zeros(count, start, max_refine, rank, message_stamp, &result)
         sig_off()
         returnvalue = []
         for i in range(result.size()):
@@ -380,7 +416,7 @@
 
     cdef c_Complex __hardy_z_function(self,c_Complex s):
         raise NotImplementedError
-    
+
     cdef int __compute_rank(self):
         raise NotImplementedError
 
@@ -390,7 +426,7 @@
     cdef void __find_zeros_v(self,double T1, double T2, double stepsize, doublevec *result):
         raise NotImplementedError
 
-    cdef void __find_zeros_via_N_v(self, long count,int do_negative,double max_refine, int rank, int test_explicit_formula, doublevec *result):
+    cdef int __find_zeros(self, long count, long start, double max_refine, int rank, const char* message_stamp, doublevec *result):
         raise NotImplementedError
 
 ##############################################################################
@@ -473,7 +509,7 @@
 
     cdef inline c_Complex __value(self,c_Complex s,int derivative):
         return (<c_Lfunction_I *>(self.thisptr)).value(s, derivative, "pure")
-            
+
     cdef inline c_Complex __hardy_z_function(self,c_Complex s):
         return (<c_Lfunction_I *>(self.thisptr)).value(s, 0, "rotated pure")
 
@@ -486,8 +522,8 @@
     cdef double __typedN(self, double T):
         return (<c_Lfunction_I *>self.thisptr).N(T)
 
-    cdef void __find_zeros_via_N_v(self, long count,int do_negative,double max_refine, int rank, int test_explicit_formula, doublevec *result):
-        (<c_Lfunction_I *>self.thisptr).find_zeros_via_N_v(count, do_negative, max_refine, rank, test_explicit_formula, result[0])
+    cdef int __find_zeros(self, long count, long start, double max_refine, int rank, const char* message_stamp, doublevec *result):
+        (<c_Lfunction_I *>self.thisptr).find_zeros(count, start, max_refine, rank, message_stamp, result)
 
     # debug tools
     def _print_data_to_standard_output(self):
@@ -500,7 +536,7 @@
             sage: from sage.libs.lcalc.lcalc_Lfunction import *
             sage: chi = DirichletGroup(5)[2] #This is a quadratic character
             sage: L=Lfunction_from_character(chi, type="int")
-            sage: L._print_data_to_standard_output() # tol 1e-15
+            sage: L._print_data_to_standard_output() # tol 1e-8
             -----------------------------------------------
             <BLANKLINE>
             Name of L_function:
@@ -624,8 +660,8 @@
     cdef double __typedN(self, double T):
         return (<c_Lfunction_D *>self.thisptr).N(T)
 
-    cdef void __find_zeros_via_N_v(self, long count,int do_negative,double max_refine, int rank, int test_explicit_formula, doublevec *result):
-        (<c_Lfunction_D *>self.thisptr).find_zeros_via_N_v(count, do_negative, max_refine, rank, test_explicit_formula, result[0])
+    cdef int __find_zeros(self, long count, long start,double max_refine, int rank, const char* message_stamp, doublevec *result):
+        (<c_Lfunction_D *>self.thisptr).find_zeros(count, start, max_refine, rank, message_stamp, result)
 
     # debug tools
     def _print_data_to_standard_output(self):
@@ -638,7 +674,7 @@
             sage: from sage.libs.lcalc.lcalc_Lfunction import *
             sage: chi = DirichletGroup(5)[2] #This is a quadratic character
             sage: L=Lfunction_from_character(chi, type="double")
-            sage: L._print_data_to_standard_output() # tol 1e-15
+            sage: L._print_data_to_standard_output() # tol 1e-8
             -----------------------------------------------
             <BLANKLINE>
             Name of L_function:
@@ -769,8 +805,8 @@
     cdef double __typedN(self, double T):
         return (<c_Lfunction_C *>self.thisptr).N(T)
 
-    cdef void __find_zeros_via_N_v(self, long count,int do_negative,double max_refine, int rank, int test_explicit_formula, doublevec *result):
-        (<c_Lfunction_C *>self.thisptr).find_zeros_via_N_v(count, do_negative, max_refine, rank, test_explicit_formula, result[0])
+    cdef int __find_zeros(self, long count, long start, double max_refine, int rank, const char* message_stamp, doublevec *result):
+        (<c_Lfunction_C *>self.thisptr).find_zeros(count, start, max_refine, rank, message_stamp, result)
 
     # debug tools
     def _print_data_to_standard_output(self):
@@ -783,7 +819,7 @@
             sage: from sage.libs.lcalc.lcalc_Lfunction import *
             sage: chi = DirichletGroup(5)[1]
             sage: L=Lfunction_from_character(chi, type="complex")
-            sage: L._print_data_to_standard_output() # tol 1e-15
+            sage: L._print_data_to_standard_output() # tol 1e-8
             -----------------------------------------------
             <BLANKLINE>
             Name of L_function:
@@ -854,8 +890,8 @@
     cdef double __typedN(self, double T):
         return (<c_Lfunction_Zeta *>self.thisptr).N(T)
 
-    cdef void __find_zeros_via_N_v(self, long count,int do_negative,double max_refine, int rank, int test_explicit_formula, doublevec *result):
-        (<c_Lfunction_Zeta *>self.thisptr).find_zeros_via_N_v(count, do_negative, max_refine, rank, test_explicit_formula, result[0])
+    cdef int __find_zeros(self, long count, long start, double max_refine, int rank, const char* message_stamp, doublevec *result):
+        (<c_Lfunction_Zeta *>self.thisptr).find_zeros(count, start, max_refine, rank, message_stamp, result)
 
     def __dealloc__(self):
         """
@@ -869,7 +905,7 @@
 
 def Lfunction_from_character(chi, type="complex"):
     """
-    Given a primitive Dirichlet character, this function returns 
+    Given a primitive Dirichlet character, this function returns
     an lcalc L-function object for the L-function of the character.
 
     INPUT:
@@ -950,10 +986,11 @@
         sage: L = Lfunction_from_elliptic_curve(EllipticCurve('37'))
         sage: L
         L-function with real Dirichlet coefficients
-        sage: L.value(0.5).abs() < 1e-15   # "noisy" zero on some platforms (see #9615)
+        sage: L.value(0.5).abs() < 1e-8
+        True
+        sage: (L.value(0.5, derivative=1) - 0.305999773835200).abs() < 1e-6
         True
-        sage: L.value(0.5, derivative=1)
-        0.305999...
+
     """
     import sage.libs.lcalc.lcalc_Lfunction
     Q = RRR(E.conductor()).sqrt() / RRR(2 * pi)
--- a/sage/src/sage/lfunctions/lcalc.py
+++ b/sage/src/sage/lfunctions/lcalc.py
@@ -225,19 +225,54 @@
         EXAMPLES::
 
             sage: I = CC.0
-            sage: lcalc.values_along_line(0.5, 0.5+20*I, 5)
-            [(0.500000000, -1.46035451), (0.500000000 + 4.00000000*I, 0.606783764 + 0.0911121400*I), (0.500000000 + 8.00000000*I, 1.24161511 + 0.360047588*I), (0.500000000 + 12.0000000*I, 1.01593665 - 0.745112472*I), (0.500000000 + 16.0000000*I, 0.938545408 + 1.21658782*I)]
+            sage: values = lcalc.values_along_line(0.5, 0.5+20*I, 5)
+            sage: values[0][0] # abs tol 1e-8
+            0.5
+            sage: values[0][1] # abs tol 1e-8
+            -1.46035451 + 0.0*I
+            sage: values[1][0] # abs tol 1e-8
+            0.5 + 4.0*I
+            sage: values[1][1] # abs tol 1e-8
+            0.606783764 + 0.0911121400*I
+            sage: values[2][0] # abs tol 1e-8
+            0.5 + 8.0*I
+            sage: values[2][1] # abs tol 1e-8
+            1.24161511 + 0.360047588*I
+            sage: values[3][0] # abs tol 1e-8
+            0.5 + 12.0*I
+            sage: values[3][1] # abs tol 1e-8
+            1.01593665 - 0.745112472*I
+            sage: values[4][0] # abs tol 1e-8
+            0.5 + 16.0*I
+            sage: values[4][1] # abs tol 1e-8
+            0.938545408 + 1.21658782*I
 
         Sometimes warnings are printed (by lcalc) when this command is
         run::
 
             sage: E = EllipticCurve('389a')
-            sage: E.lseries().values_along_line(0.5, 3, 5)
-            [(0.000000000, 0.209951303),
-             (0.500000000, -...e-16),
-             (1.00000000, 0.133768433),
-             (1.50000000, 0.360092864),
-             (2.00000000, 0.552975867)]
+            sage: values = E.lseries().values_along_line(0.5, 3, 5)
+            sage: values[0][0] # abs tol 1e-8
+            0.0
+            sage: values[0][1] # abs tol 1e-8
+            0.209951303  + 0.0*I
+            sage: values[1][0] # abs tol 1e-8
+            0.5
+            sage: values[1][1] # abs tol 1e-8
+            0.0  + 0.0*I
+            sage: values[2][0] # abs tol 1e-8
+            1.0
+            sage: values[2][1] # abs tol 1e-8
+            0.133768433 - 0.0*I
+            sage: values[3][0] # abs tol 1e-8
+            1.5
+            sage: values[3][1] # abs tol 1e-8
+            0.360092864 - 0.0*I
+            sage: values[4][0] # abs tol 1e-8
+            2.0
+            sage: values[4][1] # abs tol 1e-8
+            0.552975867 + 0.0*I
+
         """
         L = self._compute_L(L)
         CC = sage.rings.all.ComplexField(prec)
@@ -281,8 +316,31 @@
 
         EXAMPLES::
 
-            sage: lcalc.twist_values(0.5, -10, 10)
-            [(-8, 1.10042141), (-7, 1.14658567), (-4, 0.667691457), (-3, 0.480867558), (5, 0.231750947), (8, 0.373691713)]
+            sage: values = lcalc.twist_values(0.5, -10, 10)
+            sage: values[0][0]
+            -8
+            sage: values[0][1] # abs tol 1e-8
+            1.10042141 + 0.0*I
+            sage: values[1][0]
+            -7
+            sage: values[1][1] # abs tol 1e-8
+            1.14658567 + 0.0*I
+            sage: values[2][0]
+            -4
+            sage: values[2][1] # abs tol 1e-8
+            0.667691457 + 0.0*I
+            sage: values[3][0]
+            -3
+            sage: values[3][1] # abs tol 1e-8
+            0.480867558 + 0.0*I
+            sage: values[4][0]
+            5
+            sage: values[4][1] # abs tol 1e-8
+            0.231750947 + 0.0*I
+            sage: values[5][0]
+            8
+            sage: values[5][1] # abs tol 1e-8
+            0.373691713 + 0.0*I
         """
         L = self._compute_L(L)
         CC = sage.rings.all.ComplexField(prec)
--- a/sage/src/sage/schemes/elliptic_curves/lseries_ell.py
+++ b/sage/src/sage/schemes/elliptic_curves/lseries_ell.py
@@ -400,8 +400,22 @@
 
             sage: E = EllipticCurve('37a')
             sage: vals = E.lseries().twist_values(1, -12, -4)
-            sage: vals  # abs tol 1e-15
-            [(-11, 1.47824342), (-8, 8.9590946e-18), (-7, 1.85307619), (-4, 2.45138938)]
+            sage: vals[0][0]
+            -11
+            sage: vals[0][1] # abs tol 1e-8
+            1.47824342 + 0.0*I
+            sage: vals[1][0]
+            -8
+            sage: vals[1][1] # abs tol 1e-8
+            0.0 + 0.0*I
+            sage: vals[2][0]
+            -7
+            sage: vals[2][1] # abs tol 1e-8
+            1.85307619 + 0.0*I
+            sage: vals[3][0]
+            -4
+            sage: vals[3][1] # abs tol 1e-8
+            2.45138938 + 0.0*I
             sage: F = E.quadratic_twist(-8)
             sage: F.rank()
             1
--- a/sage/src/sage/modular/dirichlet.py
+++ b/sage/src/sage/modular/dirichlet.py
@@ -751,8 +751,8 @@
             sage: a = a.primitive_character()
             sage: L = a.lfunction(algorithm='lcalc'); L
             L-function with complex Dirichlet coefficients
-            sage: L.value(4)  # abs tol 1e-14
-            0.988944551741105 - 5.16608739123418e-18*I
+            sage: L.value(4)  # abs tol 1e-8
+            0.988944551741105 + 0.0*I
         """
         if algorithm is None:
             algorithm = 'pari'
--- a/sage/src/sage/schemes/elliptic_curves/ell_rational_field.py
+++ b/sage/src/sage/schemes/elliptic_curves/ell_rational_field.py
@@ -1516,7 +1516,7 @@
             sage: EllipticCurve([1234567,89101112]).analytic_rank(algorithm='rubinstein')
             Traceback (most recent call last):
             ...
-            RuntimeError: unable to compute analytic rank using rubinstein algorithm (unable to convert ' 6.19283e+19 and is too large' to an integer)
+            RuntimeError: unable to compute analytic rank using rubinstein algorithm (unable to convert ' 6.19283... and is too large' to an integer)
             sage: EllipticCurve([1234567,89101112]).analytic_rank(algorithm='sympow')
             Traceback (most recent call last):
             ...
--- a/sage/src/sage/lfunctions/zero_sums.pyx
+++ b/sage/src/sage/lfunctions/zero_sums.pyx
@@ -829,8 +829,11 @@
         EXAMPLES::
 
             sage: E = EllipticCurve("11a")
-            sage: E.lseries().zeros(2)
-            [6.36261389, 8.60353962]
+            sage: zeros = E.lseries().zeros(2)
+            sage: zeros[0] # abs tol 1e-8
+            6.36261389
+            sage: zeros[1] # abs tol 1e-8
+            8.60353962
 
         E is a rank zero curve; the lowest zero has imaginary part ~6.36. The
         zero sum with tau=0 indicates that there are no zeros at the central
